x-pack/plugins/snapshot_restore/server/lib/repository_serialization.ts

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { camelCase, snakeCase } from 'lodash';
+
+interface RepositorySettings {
+  [key: string]: any;
+}
+
+const booleanizeValue = (value: any) => {
+  if (value === 'true') {
+    return true;
+  } else if (value === 'false') {
+    return false;
+  }
+  return value;
+};
+
+const flatten = (source: any, path: any[] = []): any => {
+  if (!(source instanceof Object)) {
+    return {
+      [path.join('.')]: source,
+    };
+  }
+
+  return Object.keys(source).reduce((result, key) => {
+    const flattened: any = flatten(source[key], [...path, key]);
+    return {
+      ...result,
+      ...flattened,
+    };
+  }, {});
+};
+
+export const deserializeRepositorySettings = (settings: RepositorySettings): RepositorySettings => {
+  const flattenedSettings: RepositorySettings = flatten(settings);
```

Comment:
Maybe add a comment here? To explain that the reason we do this is that some settings, e.g. `security.principle` are returned as nested objects (`{ security: { principal: 'some_value' } }`), and that flattening the paths to these values makes it easier for the UI to render them.